### DISCUSSION TITLE
Access the raw input (formData) in addition to parsedInput

### DISCUSSION THREAD

**Discussion by Thaano** [next-safe-action/next-safe-action#311] [Q&A] [answered]
Hi,
Would it be possible to access the "raw" input in addition to the parsedInput?

I've built a form builder that allows creating forms directly in a CMS, which automatically generates a form page. My server action then receives the user's formData, fetches the corresponding form fields from the CMS, and creates a dynamic Zod validation schema.

I started moving my server actions to next-safe-action, but as far as I understand, it’s not possible to generate a dynamic Zod validation schema at runtime. Because of this, I plan to continue generating and validating the forms directly in the server action code. However, to achieve this, I need access to the raw input of the submitted form—meaning the entire formData, not just the fields already validated by the Zod schema, which cannot anticipate all the fields that might be submitted.

Here is an example of what I would need:

```javascript
const accessFormSchema = zfd.formData(
  z.object({
    birthdate: z.string(),
    file_ref: z.string(),
    form_id: z.coerce.number(),
  })
)

export type accessFormSchemaVE = z.inferFlattenedErrors<typeof accessFormSchema>['fieldErrors']

export const accessForm = publicAction
  .metadata({ actionName: 'accessForm' })
  .schema(accessFormSchema)
  .action(async ({ parsedInput, rawInput }) => {
    const { birthdate, file_ref, form_id } = parsedInput
// generate zod schema here and validate form fields from rawInput 
  })
``` 

Thanks in advance for your help!

**Comment by TheEdoRan** [comment, marked as answer]
Hey, I released [v7.10.2](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.10.2) which adds `clientInput` and `bindArgsClientInputs` props to the server code function callback object argument. You should be able to access raw inputs passed from the client inside of it from now on.

**Reply by Thaano** [reply]
Amazing! Thank you so much for your responsiveness and the solution you provided.

It turns out that I found an alternative solution to my problem in the meantime, but your release would have solved it perfectly.

Here’s what I’ve done:
```javascript
const submitFormSchema = zfd.formData(
  z.intersection(
    z.object({
      form_id: z.string(),
      data_collection_agreement: z.literal('true', {
        errorMap: () => ({
          message: 'Vous devez accepter la collecte des données.',
        }),
      }),
    }),
    z.record(z.any())
  )
)
```
Using `z.record(z.any())` allowed me to capture all the formData inputs in the parsedInput argument.

**Reply by TheEdoRan** [reply]
Glad you found a way to solve it! 😊
